### ISSUE TITLE
General

### ISSUE THREAD

**Issue by MrOCR** [mittagessen/kraken#97]
This issue is for discussing general issues, improvements, etc....

**Comment by MrOCR** [comment]
@mittagessen Does Kraken use fp32 (floating point)?
What are your thoughts  on `Mixed-Precision training`, its basically:
- fp16 models with fp32 accuracy
- Use 50% less memory
- 2x faster training, but when using Tensor Cores it might have 8x higher throughput.

https://github.com/NVIDIA/apex

https://devblogs.nvidia.com/mixed-precision-training-deep-neural-networks/
https://docs.nvidia.com/deeplearning/sdk/mixed-precision-training/index.html#pytorch
https://www.youtube.com/watch?v=i1fIBtdhjIg
http://on-demand.gputechconf.com/gtc/2018/presentation/s8923-training-neural-networks-with-mixed-precision-theory-and-practice.pdf
https://arxiv.org/abs/1710.03740

**Comment by mittagessen** [comment]
I ran some short tests using apex a while ago but the speedup wasn't that impressive. I am going to revisit the issue after I implement batching as that is the slowest part when training on GPU right now.

**Comment by Usamamalik007** [comment]
Is there some way to train data using ketos train on Google colab?

**Comment by mittagessen** [comment]
Install miniconda in the notebook and then run:

```
! wget https://raw.githubusercontent.com/mittagessen/kraken/master/environment_cuda.yml
! conda env create -f environment_cuda.yml
! conda activate kraken
```

and you should be golden. Mind you I've never used colab but that should be the general gist of it.

**Comment by MrOCR** [comment]
@mittagessen Does multiple GPUs scale-up with Kraken and seg?
So is it better to buy Two gtx 1070 ti cards or a Single rtx 2070? for training Kraken and seg.

**Comment by mittagessen** [comment]
kraken uses a miniscule amount of memory on the GPU so until I merge the batch processing patches I'm evaluating right now it doesn't matter at all what kind of GPU you use (the bottleneck is at other places). For the segmenter a card with >6Gb memory is recommended but training doesn't accelerate across multiple GPUs either.

**Comment by amitdo** [comment]
Do you plan to support quantization of the model (float to int)?

**Comment by mittagessen** [comment]
I was evaluating half-precision a while ago but the performance improvement was marginal and it felt like such a change should be really thoroughly evaluated. 8bit integer quantization isn't supported in pytorch [yet](https://github.com/pytorch/pytorch/issues/11348) although adding it outside of the framework isn't that hard. I would probably consider adding it with the new segmentation code as that is really rather slow on CPU and would benefit most of it.

For the recognizer itself the baseline normalizer is currently the computationally most intensive part of the whole pipeline, being even slower than the neural network on the CPU. Unfortunately, the summarizing LSTM layer that tesseract uses always produced worse results in my tests and I haven't found a nice replacement, yet.

So my current priority list in improving performance (cycles) is batching (GPU will be quite a bit faster), replacing the normalizer, quantization (FP16/int8), and finally fully convolutional sequence classification networks. But frankly, extracting sufficient training data of highly motivated ~~indentured servants~~ grad students for the new segmenter is a higher priority right now than improving recognition speed.

**Comment by amitdo** [comment]
https://github.com/pytorch/FBGEMM

>FBGEMM is used as a backend of Caffe2 quantized operators for x86 machines (https://github.com/pytorch/pytorch/tree/master/caffe2/quantization/server). We also plan to integrate FBGEMM into PyTorch.

**Comment by MrOCR** [comment]
@mittagessen
Clearly, batching is more important than fp16.

Note that you would need **Tensor Cores** to fully benefit from half-precision. Also note that the gtx 10x series are mostly optimized for fp32.

https://github.com/NVIDIA/apex/issues/76
https://en.wikipedia.org/wiki/List_of_Nvidia_graphics_processing_units#GeForce_20_series

**Comment by amitdo** [comment]
>Unfortunately, the summarizing LSTM layer that tesseract uses always produced worse results in my tests and I haven't found a nice replacement, yet.

Did you test tesseract itself, or use pytorch with similar techmique?

If you use pytorch - FYI, AFAIK Tesseract always scales the line height to 48 pixels.

**Comment by mittagessen** [comment]
> Did you test tesseract itself, or use pytorch with similar technique?

I more or less reimplemented the VGSL language in pytorch with the `S` layer from the original STREET implementation in tensorflow that never made it to tesseract. There are some extensions to be able to encode old ocropy models in the new format and a few new layer types (GRUs, 2D dropout). For the tests I just defined this:

```
[1,48,0,1 Cr3,3,16 Mp3,3 Lfys64 Lbx128 Lbx256 Do]
```

network in kraken which is fairly similar to what tesseract does (ReLU instead of tanh activation and they run the bidirectional LSTM layers in sequence).

> FYI, AFAIK Tesseract always scales the line height to 48 pixels.

That depends on the VGSL definition of the network. The nets in tessdata were trained either with 36 or 48 tall inputs. With the archaic texts we are working on anything below 48 is pushing it, especially if the binarization is less than perfect.

**Comment by amitdo** [comment]
Thanks for the detailed reply.

I was wrong about the 'always 48'.

>https://github.com/tesseract-ocr/tesseract/commit/f24ef67df41d36ce9238a1fc8191d18e7d0e3a76

`48` is the maximum value for line height, but the value can be less than 48 as you said.

**Comment by theceday** [comment]
This is not exactly a kraken issue but i guess it might suffer from the very same issue.
I am using following degradations but some of the generated images are very bad.
I have logged some samples during training if the loss is much higher than the current average, this is how I detected them.

Degradations are actually mix of all avaiable options. Text is currently random.

I guess I need to log every "random" variable to detect what causes this issue.
It is probably due to "hi" values but I am not very sure. It is not possible to replicate the issue with just these numbers.

"wrong" samples are not just a couple, so it needs a fix.
It might take some time to investigate the issue.

What kind of degradation numbers are good with your experience?


        elif degradationsStr=="mix":
            deglist = """
            0.5 0.0   0.5 0.0
            0.5 0.0   0.5 0.0
            0.5 0.0   0.5 0.0
            0.5 0.0   0.5 0.05
            1.0 0.3   0.4 0.05
            1.0 0.3   0.5 0.05
            1.0 0.3   0.6 0.05
            1.0 0.3   0.6 0.05
            1.0 0.3   0.4 0.1
            1.0 0.3   0.5 0.1
            1.0 0.3   0.6 0.1
            1.3 0.3   0.4 0.1
            1.3 0.3   0.5 0.1
            1.3 0.3   0.6 0.1
            """
![011770 bin](https://user-images.githubusercontent.com/26809484/50391902-92089d80-075a-11e9-8295-92a945fa34f0.png)
![019192 bin](https://user-images.githubusercontent.com/26809484/50391903-92089d80-075a-11e9-92cf-2c147f7b9744.png)

**Comment by theceday** [comment]
btw, with "hi" values disabled, validation accuracy is nearly same as training, and I am not able to "catch" a bad image, fyi. (all of them are false positive)

**Comment by amitdo** [comment]
@mittagessen,

What about the INT8 quantization we talked about a few years ago? It's now part of the framework.

https://pytorch.org/docs/stable/quantization.html

Although this feature first appeared in v1.3.0, the documentation still says:

>Quantization is in beta and subject to change.

**Comment by mittagessen** [comment]
> > Quantization is in beta and subject to change.

@amitdo The actual issue isn't so much that the API might change but that 8bit quantization is hardware dependent. You can't have a model that works on both CPU and GPU. Half-precision has been in all training paths for a while now though and doesn't change anything on the model serialization side.
